File: app/src/main/java/com/example/farith/dailynotes/MainActivity.java
package com.example.farith.dailynotes;

import android.content.Intent;
import android.database.sqlite.SQLiteDatabase;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.farith.dailynotes.ModelClass.NotesDatabaseList;
import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;


public class MainActivity extends AppCompatActivity implements View.OnClickListener, LongPressActionListener, SearchView.OnQueryTextListener {
    private static final String TAG = MainActivity.class.getSimpleName();
    private RecyclerView notesRecyclerView;
    private FloatingActionButton addButton;
    ArrayList<NotesDatabaseList> notesDatabaseData = new ArrayList<>();
    SQLiteDatabase database;
    NoteDb dbReference;
    NotesAdapter adapter;
    TextView emptyText;
    Boolean isGrid = true;
    public static final String NOTIFICATION = "com.example.farith.dailynotes.MainActivity";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        initView();
        dbReference = new NoteDb(MainActivity.this);
        dbReference.insertStatusData(database, "true");
        addButton.setOnClickListener(this);

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.convert_to_grid, menu);
        MenuItem search = menu.findItem(R.id.menu_search);
        SearchView searchView = (SearchView) search.getActionView();
        if (searchView != null)
            searchView.setOnQueryTextListener(this);
        return super.onCreateOptionsMenu(menu);
    }

//    @Override
//    public boolean onPrepareOptionsMenu(Menu menu) {
//        super.onPrepareOptionsMenu(menu);
//        if (isGrid) {
//            menu.findItem(R.id.change_to_list).setVisible(true);
//            menu.findItem(R.id.change_to_grid).setVisible(false);
//        } else {
//            menu.findItem(R.id.change_to_list).setVisible(false);
//            menu.findItem(R.id.change_to_grid).setVisible(true);
//        }
//        return true;
//    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        super.onOptionsItemSelected(item);
        if (item.getItemId() == R.id.change_to_list) {
            isGrid = false;
            dbReference = new NoteDb(MainActivity.this);
            notesDatabaseData = dbReference.readValues(database);
            Log.d(TAG, "onOptionsItemSelected:readvalues from db " + notesDatabaseData.get(0).getNotificationID());
            dbReference.updateStatusData(database, isGrid.toString());
            adapter = new NotesAdapter(MainActivity.this, notesDatabaseData);
            notesRecyclerView.setLayoutManager(new LinearLayoutManager(this));

        } else {
            isGrid = true;
            dbReference.updateStatusData(database, isGrid.toString());
//            invalidateOptionsMenu();
            adapter = new NotesAdapter(MainActivity.this, notesDatabaseData);
            notesRecyclerView.setLayoutManager(new GridLayoutManager(this, 2, LinearLayoutManager.VERTICAL, false));
        }
        notesRecyclerView.setAdapter(adapter);
        if (notesDatabaseData.size() == 0) {
            emptyText.setVisibility(View.VISIBLE);
        }
        return true;
    }

    @Override
    protected void onStart() {
        super.onStart();
        notesDatabaseData = dbReference.readValues(database);
        String data = dbReference.readStatusData(database);
        if (data.equals("true")) {
            adapter = new NotesAdapter(MainActivity.this, notesDatabaseData);
            notesRecyclerView.setLayoutManager(new GridLayoutManager(this, 2, LinearLayoutManager.VERTICAL, false));
            notesRecyclerView.setAdapter(adapter);
            if (notesDatabaseData.size() == 0) {
                emptyText.setVisibility(View.VISIBLE);


            }
        } else {
            adapter = new NotesAdapter(MainActivity.this, notesDatabaseData);
            notesRecyclerView.setLayoutManager(new LinearLayoutManager(this));
            notesRecyclerView.setAdapter(adapter);
            if (notesDatabaseData.size() == 0) {
                emptyText.setVisibility(View.VISIBLE);
            }
        }
    }

    private void initView() {
        notesRecyclerView = findViewById(R.id.notes_recyclerview);
        addButton = findViewById(R.id.add_button);
        emptyText = findViewById(R.id.emptyText);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.add_button:
                Intent noteActivityIntent = new Intent(MainActivity.this, NoteActivity.class);
                startActivity(noteActivityIntent);
        }
    }

    private String getCurrentTime() {
        DateFormat df = new SimpleDateFormat("EEE, d MMM yyyy, h:mm a");
        String date = df.format(Calendar.getInstance().getTime());
        Log.d(TAG, "getCurrentTime: date " + date);
        return date;
    }

    @Override
    public void updateRecyclerView(int position, String notificationId) {
        notesDatabaseData.remove(position);
        adapter.notifyItemRemoved(position);
        //Cancelling the notification when the item is deleted from the list
        Intent cancelNotificationIntent = new Intent(this, CancelNotification.class);
        cancelNotificationIntent.putExtra("cancel_noti", notificationId);
        sendBroadcast(cancelNotificationIntent);
        if (notesDatabaseData.size() == 0) {
            emptyText.setVisibility(View.VISIBLE);
            notesDatabaseData.clear();
        }
    }

    @Override
    public boolean onQueryTextSubmit(String s) {
        return false;
    }

    @Override
    public boolean onQueryTextChange(String s) {
        ArrayList<NotesDatabaseList> tempList = new ArrayList<>();
        for (int i = 0; i < notesDatabaseData.size(); i++) {
            if (notesDatabaseData.get(i).getNotes().toLowerCase().contains(s.toLowerCase())) {
                tempList.add(notesDatabaseData.get(i));
            }
        }
        adapter.updateList(tempList);
        return true;
    }
}

File: app/src/main/java/com/example/farith/dailynotes/ui/MainActivity.kt
package com.example.farith.dailynotes.ui

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.farith.dailynotes.R
import com.example.farith.dailynotes.databinding.ActivityMainBinding
import com.example.farith.dailynotes.modelClass.NoteClass
import com.example.farith.dailynotes.ui.adapters.NotesAdapter
import com.example.farith.dailynotes.ui.viewModels.MainViewModel
import java.util.*
import kotlin.collections.ArrayList

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private lateinit var notesAdapter: NotesAdapter
    private lateinit var noteList: List<NoteClass>
    private lateinit var viewModel: MainViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        bindData()
        setUpViewModel()
    }

    private fun bindData() {
        notesAdapter = NotesAdapter( ArrayList())
        with(binding.notesRecyclerview) {
            this.layoutManager = GridLayoutManager(
                this@MainActivity,
                2, LinearLayoutManager.VERTICAL, false
            ).also { adapter = notesAdapter }
        }
        binding.addButton.visibility = View.VISIBLE
        binding.addButton.setOnClickListener {
            val noteActivityIntent = Intent(this@MainActivity, NoteActivity::class.java)
            noteActivityIntent.putExtra("ADD_BUTTON",true)
            startActivity(noteActivityIntent)
        }
    }

    private fun setUpViewModel() {
        viewModel = ViewModelProvider(this)[MainViewModel::class.java]
        viewModel.getNoteLiveData().observe(this, { data ->
            data?.let {
                if (data.isEmpty()) {
                    binding.emptyText.visibility = View.VISIBLE
                } else {
                    notesAdapter.updateList(data)
                    notesAdapter.notifyDataSetChanged()
                }
            }
        })
    }

    override fun onStart() {
        super.onStart()
        if (::viewModel.isInitialized)
            viewModel.getNotesList()

    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.convert_to_grid, menu)
        val search = menu.findItem(R.id.menu_search)
        val searchView = search.actionView as SearchView
        searchView.setOnQueryTextListener(onQueryTextListener)
        return super.onCreateOptionsMenu(menu)
    }

    private val onQueryTextListener = object : SearchView.OnQueryTextListener {
        override fun onQueryTextSubmit(query: String?): Boolean {
            return false
        }

        override fun onQueryTextChange(newText: String?): Boolean {
            val tempList = java.util.ArrayList<NoteClass>()
            if (::noteList.isInitialized && newText != null) {
                for (i in noteList.indices) {
                    val note = noteList[i].noteContent?.lowercase()
                    if (note.equals(newText)) {
                        tempList.add(noteList[i])
                    }
                }
            }
            notesAdapter.updateList(tempList)
            return true
        }

    }


}